src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Android.cs

```diff
@@ -0,0 +1,390 @@
+                public int length;
+                public int unused; // Was raw GMT offset; always 0 since tzdata2014f (L).
+            }
+
+            private string[]? _ids;
+            private int[]? _byteOffsets;
+            private int[]? _lengths;
+
+            public AndroidTzData()
+            {
+                ReadHeader(GetTimeZoneDirectory() + _timeZoneFileName);
+            }
+
+            private unsafe void ReadHeader(string tzFilePath)
+            {
+                int size = Math.Max(Marshal.SizeOf(typeof(AndroidTzDataHeader)), Marshal.SizeOf(typeof(AndroidTzDataEntry)));
+                var buffer = new byte[size];
+                AndroidTzDataHeader header = ReadAt<AndroidTzDataHeader>(tzFilePath, 0, buffer);
+
+                header.indexOffset = NetworkToHostOrder(header.indexOffset);
+                header.dataOffset = NetworkToHostOrder(header.dataOffset);
+
+                // tzdata files are expected to start with the form of "tzdata2012f\0" depending on the year of the tzdata used which is 2012 in this example
+                sbyte* s = (sbyte*)header.signature;
+                string magic = new string(s, 0, 6, Encoding.ASCII);
+
+                if (magic != "tzdata" || header.signature[11] != 0)
+                {
+                    var b = new StringBuilder();
+                    for (int i = 0; i < 12; ++i) {
+                        b.Append(' ').Append(((byte)s[i]).ToString("x2"));
+                    }
+
+                    throw new InvalidOperationException(SR.Format(SR.InvalidOperation_BadTZHeader, tzFilePath, b.ToString()));
+                }
+
+                ReadIndex(tzFilePath, header.indexOffset, header.dataOffset, buffer);
+            }
+
+            [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2087:UnrecognizedReflectionPattern",
+                Justification = "Implementation detail of Android TimeZone")]
+            private unsafe T ReadAt<T>(string tzFilePath, long position, byte[] buffer)
+                where T : struct
+            {
+                int size = Marshal.SizeOf(typeof(T));
+                if (buffer.Length < size)
+                {
+                    throw new InvalidOperationException(SR.InvalidOperation_BadBuffer);
+                }
+
+                using (FileStream fs = File.OpenRead(tzFilePath))
+                {
+                    fs.Position = position;
+                    int numBytesRead;
+                    if ((numBytesRead = fs.Read(buffer, 0, size)) < size)
+                    {
+                        throw new InvalidOperationException(SR.Format(SR.InvalidOperation_ReadTZError, tzFilePath, position, size, numBytesRead, size));
+                    }
+
+                    fixed (byte* b = buffer)
+                    {
+                        return (T)Marshal.PtrToStructure((IntPtr)b, typeof(T))!; // Is ! the right way to handle Unboxing a possibly null value. Should there be some check instead?
```

Comment:
I don't think I'd use this approach. Instead, I would follow the same pattern we are using to read the TZifHead file in TimeZoneInfo.AnyUnix.cs.

Basically, read in a chunk of data from the file into a `byte[]`, and then write explicit code to read each chunk of the array into the structure. That way we can easily support multiple formats in the future. And we don't need to use this unsafe method with pinning the buffer. Also, we can do the endian conversion at the appropriate place (when we read it from the file).

Reply:
`TZifHead` has a set structure, whereas `ReadAt` generalizes for the two structs `AndroidTzDataHeader` and `AndroidTzDataEntry`. Would it be preferred to keep it as one function that generalizes (support multiple formats?) based on the struct, or have two separate functions to populate either a header or entry?

Reply:
I would have two separate functions, each one knowing how to read each different piece.

Reply:
Separated this function into two `LoadHeader` and `LoadEntryAt`, renamed to better describe what they are doing, and they both explicitly construct the structs. In a later change, will try to remove the struct and use out parameters instead.